### ISSUE TITLE
Allowing LocaleMessage to be Functions

### ISSUE THREAD

**Issue by philippevezina** [kazupon/vue-i18n#969]
I am trying to do something pretty complex. Here is an example that I have in French of different translations:
- Manger de la soupe
- Manger une pomme
- Manger du pain
- Manger de l'orge

As you can see, the article preceding the noun will vary based on gender and phonetics.

## Proposal
If we could have functions as a LocaleMessage, it would allow doing these more complex translations.
Example
```js
export default {
  fr: {
    sentence: function (values) {
       let article
       if (values.word.startsWith(...)) {
         article = "de l'"
       } else {
         article = 'du'
       }
       return `Manger ${article} ${value['word']}`
    }
  }
}
```

Thoughts? Comments?

**Comment by maxKimoby** [comment]
Yeah, seems like a good idea. Is it doable? Can you give a real-app example of where you can use it?

**Comment by philippevezina** [comment]
It could mostly be useful to handle complex grammar rules in some languages such as gender (masculine/feminine) that we find in Latin languages.

**Comment by prathe** [comment]
Pourquoi on est rendu à essayer de faire ça? Vu que j'ai vu ce genre de sujet échouer dans le passé, j'aurais tendance à m'en éloigner.

**Comment by exoego** [comment]
Great thanks for proposal!
This sounds worth to have for some languages.
Pull requests are welcome :prey:

**Comment by kazupon** [comment]
Hmm, 🤔 
We have to take that into account in vue-i18n-next as well.
https://github.com/intlify/vue-i18n-next/blob/882a4a1185bd95bbdb30e389301b165651fc8a01/src/core/context.ts#L44-L58

vue-i18n-next compiles the locale text messages into functions to improve performance.

code generate example: https://github.com/intlify/vue-i18n-next/blob/master/test/message/__snapshots__/compiler.test.ts.snap
runtime code: https://github.com/intlify/vue-i18n-next/blob/master/src/message/runtime.ts

**Comment by philippevezina** [comment]
@kazupon Should two PRs be done? One for this repo and on for vue-i18n-next?

**Comment by kazupon** [comment]
@philippevezina 

This proposal is good too me, because it will be fully customizable for each language.

To adopt this proposal into vue-i18n, I will try to think in a bit of function I/O specs, in order to maintain compatibility with vue-i18n-next.

In any case, this proposal was a good inspiration for the vue-i18n-next compiler like the Vue render functions.
I would like to improve the specification of the msg function generated by the vue-i18n-next compiler by including this proposal. :)

**Comment by kazupon** [comment]
@philippevezina 

I've implemented the message function.
https://github.com/kazupon/vue-i18n/pull/972

You can find the type definitions (message function I/O) of this function here.
https://github.com/kazupon/vue-i18n/pull/972/files#diff-88bec6beae84369c0376b56b3bb88fe1R9-R14

For about use case, please see this tests.
https://github.com/kazupon/vue-i18n/pull/972/files#diff-d8f3b43f12dcdbccd0af8f77c6e968c9R1-R55

Basically, I just ported from message function which is implemented in vue-i18n-next.
You can only use named formatting and list formatting.
You can't use Linked locale messages and pluralization functions because they are not parse processed in the current implementation. This is a limitation of vue-i18n v8.20 later.

Feedback, welcome :)
